dsl/flatten.cc

```diff
@@ -0,0 +1,322 @@
+        newMethodSet();
+        return classDef;
+    }
+
+    unique_ptr<ast::MethodDef> preTransformMethodDef(core::Context ctx, unique_ptr<ast::MethodDef> methodDef) {
+        auto &methods = curMethodSet();
+        if (skipMethods.contains(methodDef.get())) {
+            ENFORCE(!methods.stack.empty());
+            return methodDef;
+        }
+        int staticLevel = computeStaticLevel(methodDef.get());
+
+        // we should only move methods that /are/ nested, so if the method stack is empty, then don't bother adding
+        // space in the move queue and record the index as -1
+        if (methods.stack.size() == 0) {
+            methods.stack.emplace_back(-1, staticLevel);
+        } else {
+            methods.stack.emplace_back(methods.methods.size(), staticLevel);
+            methods.methods.emplace_back();
+        }
+
+        return methodDef;
+    }
+
+    // Returns `true` if the method is one of the modifier names in Ruby (e.g. 'private' or 'protected' or
+    // similar). This does not need to know about `module_function` because we have already re-written it in a previous
+    // DSL pass.
+    bool isMethodModifier(ast::Send &send) {
+        auto fun = send.fun;
+        return (fun == core::Names::private_() || fun == core::Names::protected_() || fun == core::Names::public_() ||
+                fun == core::Names::privateClassMethod()) &&
+               send.args.size() == 1 && ast::isa_tree<ast::MethodDef>(send.args[0].get());
+    }
+
+    // We might want to move sends as well: either if they're method modifiers like `private` or `protected` or if
+    // they're `sig`s. If so, then we'll treat them like we treat methods on our method stack
+    unique_ptr<ast::Send> preTransformSend(core::Context ctx, unique_ptr<ast::Send> send) {
+        if (send->fun == core::Names::sig() || isMethodModifier(*send)) {
+            auto &methods = curMethodSet();
+            int staticLevel = 0;
+            if (isMethodModifier(*send) && send->args.size() >= 1) {
+                // if this is a method modifier like `private` or `protected`, then we don't need to add a new scope
+                // when we traverse the method itself, so add it to the ignore set...
+                skipMethods.insert(send->args[0].get());
+                auto methodDef = ast::cast_tree<ast::MethodDef>(send->args.front().get());
+                ENFORCE(methodDef != nullptr);
+                staticLevel = computeStaticLevel(methodDef);
+            }
+
+            // we should only move methods that are nested, so if the method stack is empty, then don't bother adding
+            // space in the move queue and record the index as -1
+            if (methods.stack.size() == 0) {
+                methods.stack.emplace_back(-1, staticLevel);
+            } else {
+                methods.stack.emplace_back(methods.methods.size(), staticLevel);
+                methods.methods.emplace_back();
+            }
```

Comment:
This section means that we only move methods that _are_ nested: otherwise, they'll stay in the same place. Previously, we'd move a lot of methods that didn't need to be moved and put a lot of `emptytree` around everywhere; this lets us avoid that.